### PR TITLE
change Mod to class

### PR DESCRIPTION
**Pull request by InvisibleSymbol**
c:
(unsure if we want to type check, had it before but removed it cause the code got bulky)

### CODE REVIEW

**Comment by InvisibleSymbol** [comment]
oh yes, I forgot to change the code that actually uses it, whoops

**Comment by tybug** [comment]
Why is this useful, by the way?

I think it's useful for users to be able to get the short and long name of `replay.mods`, but this PR doesn't change anything about how we store that.

**Comment by tybug** [comment]
So after sleeping on it I'm more convinced than before that we don't want to be putting lists everywhere in init methods; it's bad enough that `Check` takes a list of `Loadables`. I'd like `Mod` to be able to represent any arbitrary, valid mod combination, instead of just a single mod.

```sh
Mod(24).short_name

File: circleguard/enums.py [hunk 1]
```diff
@@ -14,38 +14,60 @@ class Error(Enum):
...
+    4194304     : ["LastMod",       "LM"],
+    8388608     : ["TargetPractice","TP"],
+    16777216    : ["Key9",          "K9"],
+    33554432    : ["Coop",          "CO"],
+    67108864    : ["Key1",          "K1"],
+    134217728   : ["Key3",          "K3"],
+    268435456   : ["Key2",          "K2"]
+}
+        
+class Mod():
```
Comment (samuelhklumpers) [inline circleguard/enums.py]:
```
class Mod():
    @classmethod
    def init(cls):
        cls.mods = []
        
        for k, v in int_to_mod.items():
            m = cls(k)
            setattr(cls, v[1], m)
            cls.mods += [m]
    
    def __init__(self, modbits):
        self.value = modbits

    def __eq__(self, other):
        """Override the default Equals behavior"""
        return self.value == other.value

    def __ne__(self, other):
        """Override the default Unequal behavior"""
        return self.value != other.value

    def __add__(self, other):
        return Mod(self.value | other.value)

    def __hash__(self):
        return self.value

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        return " + ".join("Mod." + mod.short_name() for mod in self.decompose())

    def decompose(self):
        return [mod for mod in Mod.mods if self.value & mod.value]

    def long_name(self):
        return " ".join(int_to_mod[mod.value][0] for mod in self.decompose())

    def short_name(self):
        return "".join(int_to_mod[mod.value][1] for mod in self.decompose())
Mod.init()
```

**Comment by tybug** [comment]
I also ran into being unable to use the class in its own body, but I still very much want to provide static hints to IDEs/etc:


```python
class Mod():

    NM = None
    NF = None
    ...
    KEY2 = None



    def __init__(self, modbits):
        self.value = modbits

    def __eq__(self, other):
        """Override the default Equals behavior"""
        return self.value == other.value

    def __ne__(self, other):
        """Override the default Unequal behavior"""
        return self.value != other.value

    def __add__(self, other):
        return Mod(self.value | other.value)

    def __hash__(self):
        return self.value

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        return " + ".join("Mod." + mod.short_name() for mod in self.decompose())

    def decompose(self):
        return [Mod(mod) for mod in int_to_mod.keys() if self.value & mod]

    def long_name(self):
        return " ".join(int_to_mod[mod.value][1] for mod in self.decompose())

    def short_name(self):
        return "".join(int_to_mod[mod.value][0] for mod in self.decompose())

for k,v in int_to_mod.items():
    setattr(Mod, v[0], Mod(k))
```

and I'd also switch the int_to_mod ordering around (short first makes more sense to me):

```python
int_to_mod = {
    0           : ["NM",         "NoMod"],
    1           : ["NF",        "NoFail"],
    2           : ["EZ",          "Easy"],
    4           : ["NV",       "NoVideo"],
    8           : ["HD",        "Hidden"],
    16          : ["HR",      "HardRock"],
    32          : ["SD",   "SuddenDeath"],
    64          : ["DT",    "DoubleTime"],
    128         : ["RL",         "Relax"],
    256         : ["HT",      "HalfTime"],
    512         : ["NC",     "Nightcore"],
    1024        : ["FL",    "Flashlight"],
    2048        : ["CN",      "Autoplay"],
    4096        : ["SO",       "SpunOut"],
    8192        : ["AP",     "Autopilot"],
    16384       : ["PF",       "Perfect"],
    32768       : ["K4",          "Key4"],
    65536       : ["K5",          "Key5"],
    131072      : ["K6",          "Key6"],
    262144      : ["K7",          "Key7"],
    524288      : ["K8",          "Key8"],
    1015808     : ["KM",        "keyMod"],
    1048576     : ["FI",        "FadeIn"],
    2097152     : ["RD",        "Random"],
    4194304     : ["LM",       "LastMod"],
    8388608     : ["TP","TargetPractice"],
    16777216    : ["K9",          "Key9"],
    33554432    : ["CO",          "Coop"],
    67108864    : ["K1",          "Key1"],
    134217728   : ["K3",          "Key3"],
    268435456   : ["K2",          "Key2"]
}
```

**Comment by tybug** [comment]
short_name and long_name also work better as attributes and not functions in my opinion. Those can be set under `__init__~

**Comment by tybug** [comment]
I can't find a way to solve this without making compromises. What I'd be happy with PRing in follows:

```python

int_to_mod = {
    0           : ["NM",         "NoMod"],
    1           : ["NF",        "NoFail"],
    2           : ["EZ",          "Easy"],
    4           : ["NV",       "NoVideo"],
    8           : ["HD",        "Hidden"],
    16          : ["HR",      "HardRock"],
    32          : ["SD",   "SuddenDeath"],
    64          : ["DT",    "DoubleTime"],
    128         : ["RL",         "Relax"],
    256         : ["HT",      "HalfTime"],
    512         : ["NC",     "Nightcore"],
    1024        : ["FL",    "Flashlight"],
    2048        : ["CN",      "Autoplay"],
    4096        : ["SO",       "SpunOut"],
    8192        : ["AP",     "Autopilot"],
    16384       : ["PF",       "Perfect"],
    32768       : ["K4",          "Key4"],
    65536       : ["K5",          "Key5"],
    131072      : ["K6",          "Key6"],
    262144      : ["K7",          "Key7"],
    524288      : ["K8",          "Key8"],
    1015808     : ["KM",        "keyMod"],
    1048576     : ["FI",        "FadeIn"],
    2097152     : ["RD",        "Random"],
    4194304     : ["LM",       "LastMod"],
    8388608     : ["TP","TargetPractice"],
    16777216    : ["K9",          "Key9"],
    33554432    : ["CO",          "Coop"],
    67108864    : ["K1",          "Key1"],
    134217728   : ["K3",          "Key3"],
    268435456   : ["K2",          "Key2"]
}

class ModCombination():
    """
    An ingame osu! mod, or combination of mods.

    Notes
    -----
    This class is not meant to be instantiated. Use :class:`~.Mod` and combine
    them as necessary instead.

    A full list of mods and their specification can be found at
    https://osu.ppy.sh/help/wiki/Game_Modifiers.
    """

    def __init__(self, value):
        self.value = value
        # avoid infinite recursion with every mod decomposing into itself
        # ad infinitum
        if self.value in int_to_mod:
            self.short_name = int_to_mod[0]
            self.long_name = int_to_mod[1]
        else:
            self.short_name = "".join(int_to_mod[mod.value][0] for mod in self.decompose())
            self.long_name = " ".join(int_to_mod[mod.value][1] for mod in self.decompose())

    def __eq__(self, other):
        """Compares the ``value`` of each object"""
        return self.value == other.value

    def __ne__(self, other):
        """Compares the ``value`` of each object"""
        return self.value != other.value

    def __add__(self, other):
        """Returns a Mod representing the bitwise OR of the two Mods"""
        return ModCombination(self.value | other.value)

    def __hash__(self):
        return self.value

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        return " + ".join("Mod." + mod.short_name for mod in self.decompose())

    def __contains__(self, other):
        return bool(self.value & other.value)

    def decompose(self):
        """
        Decomposes this mod into its base component mods, which are
        :class:`~.ModCombination`\s with a ``value`` of a power of two.

        Returns
        -------
        list: :class:`~.ModCombination`
            A list of the component :class:`~.ModCombination`\s of this mod.
        """
        return [ModCombination(mod) for mod in int_to_mod.keys() if self.value & mod]

class Mod():
    """
    An ingame osu! mod.

    Notes
    -----
    A full list of mods and their specification can be found at
    https://osu.ppy.sh/help/wiki/Game_Modifiers.
    """

    NM = NoMod          =  ModCombination(0)
    NF = NoFail         =  ModCombination(1)
    EZ = Easy           =  ModCombination(2)
    NV = NoVideo        =  ModCombination(4)
    HD = Hidden         =  ModCombination(8)
    HR = HardRock       =  ModCombination(16)
    SD = SuddenDeath    =  ModCombination(32)
    DT = DoubleTime     =  ModCombination(64)
    RL = Relax          =  ModCombination(128)
    HT = HalfTime       =  ModCombination(256)
    NC = Nightcore      =  ModCombination(512)
    FL = Flashlight     =  ModCombination(1024)
    CN = Autoplay       =  ModCombination(2048)
    SO = SpunOut        =  ModCombination(4096)
    AP = Autopilot      =  ModCombination(8192)
    PF = Perfect        =  ModCombination(16384)
    K4 = Key4           =  ModCombination(32768)
    K5 = Key5           =  ModCombination(65536)
    K6 = Key6           =  ModCombination(131072)
    K7 = Key7           =  ModCombination(262144)
    K8 = Key8           =  ModCombination(524288)
    KM = keyMod         =  ModCombination(1015808)
    FI = FadeIn         =  ModCombination(1048576)
    RD = Random         =  ModCombination(2097152)
    LM = LastMod        =  ModCombination(4194304)
    TP = TargetPractice =  ModCombination(8388608)
    K9 = Key9           =  ModCombination(16777216)
    CO = Coop           =  ModCombination(33554432)
    K1 = Key1           =  ModCombination(67108864)
    K3 = Key3           =  ModCombination(134217728)
    K2 = Key2           =  ModCombination(268435456)
```

Negatives:
* mod list is basically repeated twice (`int_to_mod` and `Mod` class)
* `Mod` doesn't inherit from `Enum` as you might expect, since `Mod.NM in Mod.HR` will fail as `type(Mod.HR)` is actually `Enum`, not `ModCombination`. The only thing we sacrifice here is readability since we already implement all the magic methods Enum would otherwise take care of for us, and there is no speed impact (from limited testing).

Positives:
* full static hints provided (unlike previous code which would error on `Mod.NM in Mod.HR` since they were set to `None` and only later dynamically changed to `Mod`)
* does entirely what we want otherwise
* easily maintainable, presuming Mod values don't change on osu's side, which is unlikely but I'd still guard against if possible (it isnt here)

Comments welcome

**Pull request merged**